tuf/repository_lib.py

```diff
@@ -1535,9 +1555,235 @@ def _get_hashes_and_length_if_needed(use_length, use_hashes, full_file_path,
+    super(Leaf, self).__init__()
+    # Include the name to ensure the hash differs between elements and cannot be replayed
+    contents["name"] = name
+    self._contents = contents
+    self._name = name
+
+    if digest:
+      self._hash = digest
+    else:
+      digest_object = securesystemslib.hash.digest(algorithm=HASH_FUNCTION)
+      # Hash the canonical json form of the data to ensure consistency
+      json_contents = securesystemslib.formats.encode_canonical(contents)
+
+      digest_object.update(json_contents.encode('utf-8'))
+      self._hash = digest_object.hexdigest()
+
+  def name(self):
+    return self._name
+
+  def contents(self):
+    return self._contents
+
+  def isLeaf(self):
+    return True
+
+
+
+
+def build_merkle_tree(fileinfodict):
+  """
+  Create a Merkle tree from the snapshot fileinfo and writes it to individual snapshot files
+
+  Returns the root and leaves
+  """
+
+  # We will build the merkle tree starting with the leaf nodes. Each
+  # leaf contains snapshot information for a single metadata file.
+  leaves = []
+  nodes = []
+  for name, contents in sorted(fileinfodict.items()):
+    if name.endswith(".json"):
+      name = os.path.splitext(name)[0]
+    leaves.append(Leaf(name, contents))
+
+  # Starting with the leaves, combine pairs of nodes to build the tree.
+  # For each pair of nodes, set the first to a left child and the second
+  # as a right child. Add the resulting parent node to new_nodes. On
+  # the next iteration, pair the nodes in new_nodes. In order to handle
+  # an odd number of nodes on any iteration, if this is the last node
+  # in an odd numbered list (there is no next node), add this node to
+  # new_nodes. End the loop when there is one remaining current_node
+  # This last node will be the root of the tree.
+  current_nodes = leaves
+
+  while(len(current_nodes) > 1):
+    new_nodes = []
+    for i in range(0, len(current_nodes), 2):
+      # If there are an odd number of nodes and this is the last
+      # node, add this node to the next level.
+      if i + 1 >= len(current_nodes):
+        new_nodes.append(current_nodes[i])
+      # Otherwise, use the next two nodes to build a new node.
+      else:
+        n = InternalNode(current_nodes[i], current_nodes[i+1])
+        # Add this node to the next level, and to a list of all nodes
+        new_nodes.append(n)
+        nodes.append(n)
+    current_nodes = new_nodes
+
+  # The only node remaining in current_nodes will be the root node.
+  root = current_nodes[0]
+
+  # Return the root node and the leaves. The root hash must be used along with the
+  # path to verify the tree. The root hash should be securely sent to
+  # each client. To do so, we will add it to the timestamp metadata.
+  # The leaves will be used to find the path to each leaf and send
+  # this path to the client for verification
+  return root, leaves
+
+def write_merkle_paths(root, leaves, storage_backend, merkle_directory):
```

Comment:
```suggestion
def _write_merkle_paths(root, leaves, storage_backend, merkle_directory):
```
Internal methods should add an underscore prefix to the method name.